default/python/universe_generation/galaxy.py

```diff
@@ -1,35 +1,297 @@
+        outer = set()
+        ring = set([1])
+        while ring and i_radius <= i_radius_max:
+            inner = outer
+            outer = set(self._square_indices_containing_cell(cell, i_radius))
+            ring = outer - inner
+            yield ring
+            i_radius += 1
+
+    # All neighbors within max_dist
+    def neighbors(self, p):
+        _neighbors = []
+        i_radius_close_enough = floor(self.max_dist / self.cell_size)
+        for i, ring in enumerate(self._generate_rings_around_point(p, self.max_dist)):
+            if i <= i_radius_close_enough:
+                _neighbors += [pos for cell in ring for pos in self.grid[cell]]
+            else:
+                _neighbors += [pos for cell in ring for pos in self.grid[cell] if
+                               util.distance(pos, p) <= self.max_dist]
+        return _neighbors
+
+    # nearest neighbor at any distance
+    def nearest_neighbor(self, p):
+        for ring in self._generate_rings_around_point(p):
+            candidates = [pos for cell in ring for pos in self.grid[cell]]
+            if candidates:
+                (_, pt) = min((util.distance(pos, p), pos) for pos in candidates)
+                return [pt]
+        return []
 
-    def too_close_to_other_positions(self, x, y):
+    def too_close_to_other_positions(self, pos):
         """
         Checks if the specified position is too close to the positions stored in the grid
         """
-        cell_x = int(x / self.cell_size)
-        cell_y = int(y / self.cell_size)
-        upper_left_x = max(0, cell_x - 1)
-        upper_left_y = max(0, cell_y - 1)
-        lower_right_x = min(self.width - 1, cell_x + 1)
-        lower_right_y = min(self.width - 1, cell_y + 1)
-        return any(util.distance(pos[0], pos[1], x, y) < self.min_dist for cx in range(upper_left_x, lower_right_x + 1)
-                   for cy in range(upper_left_y, lower_right_y + 1) for pos in self.grid[cx][cy])
+        pos_cell = self.cell(pos)
+        return any(util.distance(p2, pos) < self.min_dist
+                   for p2_cell in self._square_indices_containing_cell(pos_cell, 1)
+                   for p2 in self.grid[p2_cell])
+
+
+class DSet(object):
+    """
+    A set of pos that is disjoint (not connected) to other sets.
+
+    If the set is of one item then it has no parent. parent equals None.
+    Otherwise the set points to the root/representative item of the set via parent.
+
+    A private child class of DisjointSets to store the parent and rank of a disjoint set object
+    """
+    def __init__(self, pos):
+        self.pos = pos
+
+        # stores None or the parent
+        self.parent = None
+
+        # starts at 0, never decreases and only increases
+        # when this object is the parent and two equal rank sets are merged
+        self.rank = 0
+
+    def bind_parent(self, parent):
+        assert self != parent
```

Comment:
You should change it to 'is parent' to check if it is same object or define compare method.